Sources/Imperial/Services/Auth0/Auth0Router.swift

```diff
@@ -0,0 +1,90 @@
+import Vapor
+import Foundation
+
+public class Auth0Router: FederatedServiceRouter {
+    public let baseURL: String
+    public let tokens: FederatedServiceTokens
+    public let callbackCompletion: (Request, String)throws -> (Future<ResponseEncodable>)
+    public var scope: [String] = [ ]
+    public var requiredScopes = [ "openid" ]
+    public let callbackURL: String
+    public let accessTokenURL: String
+
+    private func providerUrl(path: String) -> String {
+        return self.baseURL.finished(with: "/") + path
+    }
+    
+    public required init(callback: String, completion: @escaping (Request, String)throws -> (Future<ResponseEncodable>)) throws {
+        let auth = try Auth0Auth()
+        self.tokens = auth
+        self.baseURL = "https://\(auth.domain)"
+        self.accessTokenURL = baseURL.finished(with: "/") + "oauth/token"
+        self.callbackURL = callback
+        self.callbackCompletion = completion
+    }
+    
+    public func authURL(_ request: Request) throws -> String {
+        let path="authorize"
+        let scopes = self.scope + self.requiredScopes
+        let scopeString = scopes.joined(separator: " ").addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)!
+
+        let params=[
+            "response_type=code",
+            "client_id=\(self.tokens.clientID)",
+            "redirect_uri=\(self.callbackURL)",
+            "scope=\(scopeString)",
+//            "state=xyzABC123" // TODO: to prevent CSRF attacks
+        ]
+        let rtn = self.providerUrl(path: path + "?" + params.joined(separator: "&"))
+        return rtn
+    }
+    
+    public func fetchToken(from request: Request)throws -> Future<String> {
+        let code: String
+        if let queryCode: String = try request.query.get(at: "code") {
+            code = queryCode
+        } else if let error: String = try request.query.get(at: "error") {
+            throw Abort(.badRequest, reason: error)
+        } else {
+            throw Abort(.badRequest, reason: "Missing 'code' key in URL query")
+        }
+        
+        let body = Auth0CallbackBody(clientId: self.tokens.clientID,
+                                     clientSecret: self.tokens.clientSecret,
+                                     code: code,
+                                     redirectURI: self.callbackURL)
+        
+        return try body.encode(using: request).flatMap(to: Response.self) { request in
+            guard let url = URL(string: self.accessTokenURL) else {
+                throw Abort(.internalServerError, reason: "Unable to convert String '\(self.accessTokenURL)' to URL")
+            }
+            request.http.method = .POST
+            request.http.url = url
+            request.http.contentType = .urlEncodedForm
+
+            print("request url: \(request.http.method) \(url.absoluteString)")
```

Comment:
We shouldn't just be printing stuff in the library. If it's useful to apps then it should go through the logger

Reply:
Oops, sorry about that, I missed those prints.